tests/test_service_injection.py

```diff
@@ -0,0 +1,232 @@
+    """
+
+    @inject(alias=interfaces.Store)
+    class NotAFactoryStore:
+        """
+        A test store thats not using a factory pattern
+        """
+
+        pass
+
+    # Instanitate injector
+    inj: Injector = Injector(nop_config, {"store": NotAFactoryStore})
+    inj.configure_service(interfaces.Store, "store", {})
+
+    @inject
+    def get_me_a_non_factory_store(store: interfaces.Store) -> interfaces.Store:
+        return store
+
+    # Instantiating twice gets us exaactly the same object back
+    instantiated_store_1 = get_me_a_non_factory_store()
+    instantiated_store_2 = get_me_a_non_factory_store()
+    assert instantiated_store_1 is instantiated_store_2
+
+
+def test_is_a_factory():
+    """
+    Confirm that services where a factory is specified, the
+    service in question is a new object each time an @inject
+    decorated function is called.
+    """
+
+    @inject(alias=interfaces.Store, use_factory=True)
+    class IsAFactoryStore:
+        """
+        A test store that is using a factory pattern
+        """
+
+        pass
+
+    # Instanitate injector
+    inj: Injector = Injector(nop_config, {"store": IsAFactoryStore})
+    inj.configure_service(interfaces.Store, "store", {})
+
+    @inject
+    def get_me_a_factory_store(store: interfaces.Store) -> interfaces.Store:
+        return store
+
+    # Instantiating twice gets us different objects back
+    instantiated_store_1 = get_me_a_factory_store()
+    instantiated_store_2 = get_me_a_factory_store()
+    assert instantiated_store_1 is not instantiated_store_2
+
+
+def test_configurable_implementations():
+    """
+    Implementations can be selected using just the configuration.ini
+    """
+    create_app(sanic_test_mode=True, config=nop_config)
+
+
+# TODO - see NOTE in method body.
+@pytest.mark.skip(
+    reason="Cannot fail until we've defined our first protocol with methods"
+)
+def test_incorrect_interface_is_raised():
+    """
+    The expected exception is raised if an implementation does not
+    support the required protocols.
+    """
+
+    class WrongUn:
+        pass
+
+    with pytest.raises(ProtocolError):
+        # NOTE: When we have out first interface with methods, pass "WrongUn"
+        # to it (as currently per messenger) and remove skip decorator.
+        create_app(messenger=WrongUn, sanic_test_mode=True, config=nop_config)
+
+
+def test_bad_config_raises():
+    """
+    Specifying unknown implementations via configuration raises an appropriate
+    exception.
+    """
+
+    config = copy.deepcopy(nop_config)
+    config["STORE"]["default_implementation"] = "I'm not a thing that exists"
+
+    with pytest.raises(UnknownImplementationError):
+        create_app(sanic_test_mode=True, config=config)
+
+
+def test_service_injected_into_service():
+    """
+    Confirm that an injectable service can be injected into the
+    constructor of another injectable service.
+    """
+
+    Injector._test_only_reset_container()
```

Comment:
It may be worth adding an [auto-use test fixture](https://docs.pytest.org/en/7.0.x/how-to/fixtures.html#autouse-fixtures-fixtures-you-don-t-have-to-request) to reset Injector state between tests.